### ISSUE TITLE
Xposed and Pokémon GO 0.37.0

### ISSUE THREAD

**Issue by pperle** [ELynx/pokemon-go-xposed-mitm#24]
I installed Magisk as descriped on [XDA](http://forum.xda-developers.com/android/software/mod-magisk-v1-universal-systemless-t3432382) and I can play Pokémon GO and also load all xposed modules.

The hooks have changed, "doSyncRequest" & "readDataSteam" are no longer available, see https://github.com/chuparCh0pper/PoGoIV_xposed/issues/23 .

**Comment by ELynx** [comment]
My phone is still 4.4.2, and will not get any updates. Magisk is 5.0+, I am out of luck here.
I am thankful that you care, but looks like it is game over for me.

**Comment by cire27rn** [comment]
Thank you very much for this xposed app. The update sucks but im happy though coz i still spoof haha. Thank you once again

**Comment by ELynx** [comment]
@cire27rn Thanks for support! Glad that you can do what you want.

I managed to install Android 5.0 to my Note 3, and have Magisk now. So I can have root, which I need for stuff much more important than PoGo, _and_ PoGo. Until they ban Magisk, anyway :)
XPosed for Sams-Andy 5.0 is too big trouble for me to install, with deodexing and such. So no XPosed hacks for PoGo from me.

**Comment by mazzod** [comment]
@ELynx thanks for your work, I really appreciate you module, in particular the possibility to see how much time the lured pokestop have :smile:

**Comment by igoticecream** [comment]
@ELynx awesome mitm module.

May i use your mitm\* classes on a project I'm working on? Of course I will mention you and this repo on the credits... you dont mind, do you?

**Comment by ELynx** [comment]
@igoticecream Sure, use them hovewer you like, just remember licensing.
Whole repo is GNU GPL3, AFAIK this means that code that uses it should be definitely opensourced and probably GNUGPLed itself.
If it is possible, leave a link to your repo in Issues of my module. I will look there ince in a while, may be merge something back.

**Comment by FreedomMercenary** [comment]
@ELynx thanks for the great work, hope this will be helpful.  I've tested on 0.35 and 0.37, but if others can help that would be great.  I also have a Note 3 and can tell you that you have a couple of options should you care to want to still use:
- Downgrade to 0.35 apk and stay on that for now
- If you have root, unlock your bootloader and flash CM (6.0 / MM) build available on XDA.  That build works with systemless root and Xposed.  I have not yet tested suhide on it as that's not my primary phone.

**Comment by ELynx** [comment]
@FreedomMercenary Thank you!

I am considering CM for some time, now I am checking builds. I would like to see where SafetyNet updates are going, to see what builds will work in the end.
Only problem that Note 3 is my main/only phone, and I need reliable build (especially RIL), and I use SPen for drawing, so I am checking what builds have pressure-sense enabled. If stars align just right, I will be back in action with systemless/XPosed.

I encourage you to continue with your fork, because you have working solution now, not later. If you can make a build, I can send you signing keys and add as developer on XPosed repository so you can publish and edit. For me this was always a community project, so if it goes on after I cant I am only happy.

**Comment by pperle** [comment]
All I could ever get was "PLAYER_UPDATE" RequestType (while login) wrapping the InputStream of "getInputStream" in a RequestEnvelope.

``` java
try {
    Envelopes.RequestEnvelope requestEnvelop = Envelopes.RequestEnvelope.parseFrom(byteArrayOutputStream.toByteArray());
    XposedBridge.log("[requestEnvelop] requestId = " + requestEnvelop.getRequestId());
    XposedBridge.log("[requestEnvelop] getRequestsCount = " + requestEnvelop.getRequestsCount());
    List<Requests.RequestType> requestList = new ArrayList<Requests.RequestType>();
    for (int i = 0; i < requestEnvelop.getRequestsCount(); i++) {
        XposedBridge.log("[requestEnvelop] doSyncRequest - " + requestEnvelop.getRequests(i).getRequestType().toString());
        requestList.add(requestEnvelop.getRequests(i).getRequestType());
    }
    requestMap.put(requestEnvelop.getRequestId(), requestList);
} catch (InvalidProtocolBufferException e) {
    XposedBridge.log("[requestEnvelop] Parsing requestEnvelop failed " + e);
    return;
}
```

But I think RequestEnvelope should be in "getOutputStream", shouldn't it?
My Problem is I cannout get a byte-array of the OutputStream of "getOutputStream".

**Comment by ELynx** [comment]
@chuparCh0pper look at code I merged into branch right-now;
I think that may not work with 0.39.1 already, since there is almost nothing left of NiaNet. But it is still possible to blindly hack on every connection.

I don't think that using ...Envelope with get...Stream is a good idea. Result streams may yet have any data, or connection may have errors. That is why I added Mitm... classes.

Also may be when Java code is called from native, and not other Java, it is not hijacked by XPosed. So in case NiaNet->SomeClass->someMethod XPosed can fake someMethod, but in case "native code"->SomeClass->someMethod it cannot.
So if they moved (almost) all calls to network code to native code, nothing is going through Mitm now.

**Comment by FreedomMercenary** [comment]
It would be great for some other people to test, but this one I have been using since I pushed the code and it definitely works for 0.37 and 0.39, which I updated to this morning.  Looking at the code for the repo in @chuparCh0pper the issue I'm having is that the MITM code appears much more dependent on the request information.  The code here in @ELynx worked fine without the request ID (and didn't appear to need request type at all) as long as I identified the traffic from the location it was going to as it didn't appear to grab the request type and work off of that.

The notes about getOutputStream are what I was running into yesterday.  I've moved on from there as getting the byte array appears to be not doable there, but am working on hooking the OkHttp library that's used deeper by Niantic and likely is the best spot left to grab the outgoing data to get request info.  So far this hasn't been as easy as I would have hoped, but that's the path I'm headed down now.
